# Installing Our Own Local Library

Why do we need to do this? Because in the next set of lessons, we will use our own helper functions from the local `wdo` package.

The geospatial module lives in:

```text
src/wdo
```

This notebook walks through the install idea and organizes the assignment problems in one place.


## How to Install Our Own Lib

Our project contains reusable Python code that lives in:

```text
src/wdo
```

Instead of placing Python modules directly in the project root, we use a **`src` layout** to keep the code organized and separate from:

- notebooks
- datasets
- markdown files
- setup scripts

### Example structure

```text
.
├── data
├── notebooks
├── pyproject.toml
└── src
    └── wdo
```

Notice the `pyproject.toml` file. This is important.


## What `pyproject.toml` Does

The `pyproject.toml` file tells Python’s packaging tools:

1. the project name
2. the Python version requirements
3. where the package code lives (`src/`)

So Python knows that the real code is here:

```text
src/wdo
```


## What This Command Does

Before running notebooks or scripts, run:

```bash
pip install -e .
```

This command:

1. installs the project into the current virtual environment
2. makes the `wdo` package importable anywhere in the project
3. uses **editable mode**, so changes to the source code are immediately available


## Why Editable Mode Matters

After running:

```bash
pip install -e .
```

we can write:

```python
from wdo.geo import haversine_km
```

and Python will find the code inside:

```text
src/wdo
```

Even better, if we modify files in `src/wdo`, we do **not** need to reinstall the package.


## Quick Summary

Run this once after you clone or copy the files into your project:

```bash
pip install -e .
```

When it works, you should usually see something like:

```text
src
├── wdo
└── wdo.egg-info
```

Without this step, Python may not know where to find the `wdo` package.


## Optional: check project root from inside the notebook

This cell helps you confirm whether you are in the right project tree.


In [30]:
%cd "/Users/ricardoayala/Desktop/ricardoayala2510-Spatial-Data-Mapping/assigments completed/02-Missile_Geometry_101"
%pip install -e .

/Users/ricardoayala/Desktop/ricardoayala2510-Spatial-Data-Mapping/assigments completed/02-Missile_Geometry_101
Obtaining file:///Users/ricardoayala/Desktop/ricardoayala2510-Spatial-Data-Mapping/assigments%20completed/02-Missile_Geometry_101
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
  Building editable for wdo (pyproject.toml) ... done
  Created wheel for wdo: filename=wdo-0.0.0-0.editable-py3-none-any.whl size=1241 sha256=dec7ac03bf691a6456639b8a5627340d676f6e89555b5919bf879ac28f9a99ff
  Stored in directory: /private/var/folders/4g/brtqy4696sq_j5rk22x4zvch0000gn/T/pip-ephem-wheel-cache-x7n6ogm0/wheels/9d/47/24/36cba67d8f7629031494b92266ec05a4a2c367a07bae0a6b16
Successfully built wdo
  Attempting uninstall: wdo
    Found existing installation: wdo 0.0.0
    Uninstalling wdo-0.0.0:
      Successfully uninstalled wdo-0.0.0
No

In [31]:
from pathlib import Path

cwd = Path.cwd().resolve()
repo_root = next((p for p in [cwd, *cwd.parents] if (p / "pyproject.toml").exists()), None)

print("Current working directory:")
print(cwd)
print()
print("Detected project root:")
print(repo_root)
print()
print("Found pyproject.toml?:", repo_root is not None)

Current working directory:
/Users/ricardoayala/Desktop/ricardoayala2510-Spatial-Data-Mapping/assigments completed/02-Missile_Geometry_101

Detected project root:
/Users/ricardoayala/Desktop/ricardoayala2510-Spatial-Data-Mapping/assigments completed/02-Missile_Geometry_101

Found pyproject.toml?: True


## Assignment

### Install Library

- Run `pip install -e .` from the terminal at the same level as the `src` folder and the `pyproject.toml` file.
- Check to make sure the `wdo.egg-info` folder shows up.
- Then get the following code to run using functions from the `wdo.geo` module.


## Problem One

- Import `haversine_km`
- Calculate the distance between:

```text
[25.7853, -80.188189]
[34.046211, -118.256405]
```

- What is the distance in miles?
- What cities are we working with?

These coordinates are approximately:

- **Miami, Florida**
- **Los Angeles, California**


In [32]:
from src.wdo.geo import haversine_km

lat1, lon1 = 25.7853, -80.188189
lat2, lon2 = 34.046211, -118.256405

distance_km = haversine_km(lat1, lon1, lat2, lon2)
distance_miles = distance_km * 0.621371

print("Distance (km):", distance_km)
print("Distance (miles):", distance_miles)
print("Cities: Miami, Florida  →  Los Angeles, California")

Distance (km): 3759.1904118697207
Distance (miles): 2335.8519054139
Cities: Miami, Florida  →  Los Angeles, California


### Expected approximate answer

If your local `wdo.geo.haversine_km` matches the standard haversine formula, the result should be about:

- **3759.19 km**
- **2335.63 miles**


## Problem Two

Calculate a bounding box around the same two city points using `bbox_latlon`.


In [36]:

from src.wdo.geo import bbox_latlon, LatLon

lat1, lon1 = 25.7853, -80.188189
lat2, lon2 = 34.046211, -118.256405

points = [
    LatLon(lat1, lon1),
    LatLon(lat2, lon2)
]

bbox = bbox_latlon(points)

print("Bounding box:")
print(bbox)

Bounding box:
(25.7853, -118.256405, 34.046211, -80.188189)


### Reference min/max values

A simple min/max bounding box around the two points would be:

- south = **25.7853**
- west = **-118.256405**
- north = **34.046211**
- east = **-80.188189**

Your library may return the same values in a tuple, dict, or helper object.


## Problem Three

Using GeoJSON.io as a helper:

- create a polygon approximately the size of a small state
- create one point inside the polygon
- create one point outside the polygon
- run `point_in_polygon` to verify the results


In [37]:
from src.wdo.geo import point_in_polygon

polygon = [
    (33.0, -100.5),
    (35.0, -100.5),
    (35.0, -98.0),
    (33.0, -98.0),
]

inside_point = (34.0, -99.2)
outside_point = (36.0, -99.2)

print("Inside point result:", point_in_polygon(inside_point, polygon))
print("Outside point result:", point_in_polygon(outside_point, polygon))

Inside point result: True
Outside point result: False


### What you should see

If everything is working correctly, the output should be roughly:

```text
Inside point result: True
Outside point result: False
```


## Final Notes

- Use the terminal install first if possible.
- Restart the notebook kernel after installation.
- If an import still fails, double-check:
  - your active virtual environment
  - the folder containing `pyproject.toml`
  - whether `wdo.egg-info` was created
